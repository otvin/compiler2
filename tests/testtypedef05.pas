program testtypedef05(output);
{test passing user-defined enumerated types byval and byref to procedures and returning enumerated type from function}

type nato = (alpha, bravo, charlie, delta, echo, foxtrot, golf, hotel, india, juliett, kilo, lima, mike);

var a,b:nato;

procedure writethenato(pp:nato);
begin
	if pp = alpha then writeln('alpha');
	if pp = bravo then writeln('bravo');
	if pp = charlie then writeln('charlie');
	if pp = delta then writeln('delta');
	if pp = echo then writeln('echo');
	if pp = foxtrot then writeln('foxtrot');
	if pp = golf then writeln('golf');
	if pp = hotel then writeln('hotel');
	if pp = india then writeln('india');
	if pp = juliett then writeln('juliett');
	if pp = kilo then writeln('kilo');
	if pp = lima then writeln('lima');
	if pp = mike then writeln('mike');
end;


function doublepred(n:nato):nato;
var c:nato;
begin
	if n > alpha then
		c := pred(n)
	else
		c := n;


	if c > alpha then 
		c := pred(c);

	doublepred := c;
end;


procedure succtotwovariables(var y,z:nato);
begin
	if y < mike then 
		y := succ(y);

	if z < mike then
		z := succ(z);
end;


procedure foo(u,v:nato);
var d:nato;
begin
	write('v = ');
	writethenato(v);
	d := doublepred(v);
	write('doublepred on that d = ');
	writethenato(d);
	write('and u = ');
	writethenato(u);
	succtotwovariables(d,u);
	write('now d = ');
	writethenato(d);
	write('and now u = ');
	writethenato(u);
end;

begin
	a := echo;
	b := foxtrot;

	write('before, a and b equal ');
	writethenato(a);
	writethenato(b);
	foo(a,b);
	write('after, a and b equal ');
	writethenato(a);
	writethenato(b);

	writeln('now let''s do it again ');

	a := india;
	b := delta;

	write('before, a and b equal ');
	writethenato(a);
	writethenato(b);
	foo(a,b);
	write('after, a and b equal ');
	writethenato(a);
	writethenato(b);
end.

	
	
