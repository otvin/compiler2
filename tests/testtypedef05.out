before, a and b equal echo
foxtrot
v = foxtrot
doublepred on that d = delta
and u = echo
now d = echo
and now u = foxtrot
after, a and b equal echo
foxtrot
now let's do it again 
before, a and b equal india
delta
v = delta
doublepred on that d = bravo
and u = india
now d = charlie
and now u = juliett
after, a and b equal india
delta
